### ISSUE TITLE
More robust approach to filtering/grouping

### ISSUE THREAD

**Issue by Craksy** [jojolepro/entity_component#2]
First off, I'd just like to thank you for the awesome job you've done here.
I just recently learned about the ECS pattern, and having such a simple, straight to the point, zero-bullshit API, means that I can worry about learning the pattern rather than just reading documentation.

And so far it's also been a great foundation to build upon. It's all very transparent, and I've found it easy to extend it with my own abstractions.
My only issue is that the `join!` macro is super fragile. In fact so much that I ended up just working with the bitsets directly.
It's like the one dirty spot on an otherwise shiny surface, and it also feels a bit out of place with everything else being so transparent and extensible.

Perhaps something nice could be done with a derive macro or something, so instead of doing all your joining business "on the spot", somehow tie filtering to the type system, and make it possible to create reusable groups?
This could potentially be used not just for systems, but also as a convenient way to spawn entities with common sets of components.
I imagine it would be easier to maintain, and that it would feel more like the rest of the API.

I realize you haven't worked on this project for a while, but let me know what you think. Perhaps I'll take a crack at it when I find the time.

**Comment by AnneKitsune** [comment]
I totally agree that the join macro is the worst part of this library!
I would definitely be interested in a way to join using the type system directly.

My main goal with the join macro is for it to be as transparent as possible: you add a join!(list of component storages) and you get an iterator back. Ideally, we could make something like a regular function that takes a list of arbitrary components types and return an iterator of those types. I don't think the rust type system supports this however :thinking: 

Derive macros are an option. We could have "join structs" that define the return values of the join iterator. 
Then maybe we could use it like so:

```
#[join]
struct MyJoinGroup {
    a: Comp1,
    #[mut]
    b: Option<Comp2>,
}
...
// The join function is generated and return an iterator over & and &mut references.
for (a, b) in MyJoinGroup::join(comp1, comp2) {
}
```

I think it takes a bit too much text to define the join group struct however.. ideas?

**Comment by Craksy** [comment]
Absolutely. I think that you're right that it'd get a bit too verbose. Preferably it shouldn't be a choice between joining or convenience, right?

I was playing around with something yesterday, but unfortunately I don't have a working prototype just yet. In fact I'm not even sure if it's a viable approach.

Anyway the basic idea is to have some macro implementations for different sized tuples, that allows for stuff like flattening, getting resources from a `World` and joining bitsets.

Then perhaps it'd be possible to impl `IntoSystem` for something like `Fn(JoinGroup<(TRef1, TRef2, TRef3), (TMut1, TMut2>)`
and then use it like
```rs
fn some_system(join_group: JoinGroup<(Mesh, Shader, Texture), (Transform)>) -> SystemResult {
  for ((mesh, shader, tex), (mut tf)) in join_group{
     // do some rendering I guess
  }
}
```

Again, I'm not sure if it can even be done.
One thing I really like about it though is how similar it is to the existing API; you define the dependencies of your system in the function signature, and you get back exactly what you asked for (although, ideally join groups would implicitly wrap types in `Components<T>`. include the `&Entities` bitset and flatten options I guess?)

this way reusable groups could simply be defined using type aliases:
```rs
type PlayerGroup = JoinGroup<(Name, Faction, Level), (Health, Ammo, Transform)>;
```

I guess it could also be flattened and use a single tuple to be more in line with the current system. My initial thought was just that it'd be easier to implement all the tuple stuff if done separately for refs and muts.
